scrapy/pipelines/files.py

```diff
@@ -417,7 +417,14 @@ def _warn():
         ## end of deprecation warning block
 
         media_guid = hashlib.sha1(to_bytes(url)).hexdigest()  # change to request.url after deprecation
-        media_ext = os.path.splitext(url)[1]  # change to request.url after deprecation
+        
+        media_ext = os.path.splitext(url)[1] # change to request.url after deprecation
+        if not media_ext[1:-1].isalnum():
+            # the file extension contains non alnum characters. e.g. "?"
+            # this usually happens if url is somethin like below
+            #   https://example.net/some.jpg?ab=1e9b29b5d3d0aa13d601505c6a67ee3e&dc=59C43AAF
+            media_ext = os.path.splitext(url.split('?',1)[0])[1]
```

Comment:
why not removing the `media_ext` instead of resplitting `url` ?

Another approach that looks more natural is to use urlparse to remove the query string from url.

```python
>>> from urllib.parse import urlparse
>>> url = 'https://example.net/some.jpg?ab=1e9b29b5d3d0aa13d601505c6a67ee3e&dc=59C43AAF'
>>> os.path.splitext(urlparse(url).path)
('/some', '.jpg')
```